# 0. Intro  
## 0.1 Author Bio

**Birth** 21.08.1993 - Cremona  

**University:**
  - 2012/2013 - Brescia $\rightarrow$ $1^{st}$ year of university, Mathematics degree at Università Cattolica. After doing the exams I started to see my real way: I love mathematics and theory but it must have an application as well.  
  - 2014/2015 - Parma $\rightarrow$ $2^{nd}$ and $3^{rd}$ years, Computer Science degree at Università degli Studi di Parma
  - 08.07.2015 - Graduation
  
**Moved to Florence:**
  - October 2015
    - Started working in a consulting firm in branch of DataWarehousing, analisys and reporting
    - Started studying for the $2^{nd}$ leved degree at Università degli Studi di Firenze
      
**Back home again:**
  - February 2020
    - Moved back in Cremona (maybe in the worst moment at all)
    - Started working in Data Management Tribe at ING (Milan)

## 0.2 Environment
I've been using Ubuntu since 2014 on my computer but in January I decided to dual-boot it in order to have a small Windows environment again for particular needs.  
Techincal details:
  - CPU: Intel$^{®}$ Core™ i7-5500U CPU @ 2.40GHz × 4 
  - RAM: 8GB
  - Model: Lenovo Z70
  - OS: Ubuntu 20.04 LTS (64-bit)
  - Python Interpreter: Python

Importing some modules...

In [103]:
import json
import pandas as pd
from IPython.display import display_html

---
# 1. Familiarize with the data
Take your time to familiarize with the data, especially its tree structure and write down a quick summary (git commits history, collaborators and maintainers, what it contains and whatever you observe).

## 1.1 

---
# 2. Load data
In this part, the files of our interest are: the one at  https://github.com/pcm-dpc/COVID-19/blob/master/dati-json/dpc-covid19-ita-regioni.json and its companion at https://github.com/pcm-dpc/COVID-19/blob/master/dati-json/dpc-covid19-ita-province.json

Load both of them in two different Python objects and inspect their contents, write down your thoughts and observations of their structure. You can both use plain Python dictionaries (for a preliminary step) and Pandas dataframes (required).

## 2.1 First of all: Let's get updated data

In [2]:
covid_folder = "../COVID-19"
# If folder exists --> Get updated data, otherwise clone repo
!test -d "$covid_folder" \
    && git -C "$covid_folder" pull \
    || git -C ".." clone https://github.com/pcm-dpc/COVID-19.git
    
!mkdir -p Resources
!cp -v "$covid_folder"/dati-json/dpc-covid19-ita-regioni.json Resources
!cp -v "$covid_folder"/dati-json/dpc-covid19-ita-province.json Resources

Già aggiornato.
'../COVID-19/dati-json/dpc-covid19-ita-regioni.json' -> 'Resources/dpc-covid19-ita-regioni.json'
'../COVID-19/dati-json/dpc-covid19-ita-province.json' -> 'Resources/dpc-covid19-ita-province.json'


## 2.2 Load into Python Dataframes
### 2.2.1 City Dataframe

In [90]:
with open('Resources/dpc-covid19-ita-province.json') as f:
    d = json.load(f)
cities  = pd.DataFrame(d)    

After importing the data we can easily notice that there are some dirty data

In [94]:
display(cities)
display(cities.loc[cities.denominazione_provincia == "In fase di definizione/aggiornamento"])

data stato  codice_regione denominazione_regione  \
0      2020-02-24T18:00:00   ITA              13               Abruzzo   
1      2020-02-24T18:00:00   ITA              13               Abruzzo   
2      2020-02-24T18:00:00   ITA              13               Abruzzo   
3      2020-02-24T18:00:00   ITA              13               Abruzzo   
4      2020-02-24T18:00:00   ITA              13               Abruzzo   
...                    ...   ...             ...                   ...   
17995  2020-07-10T17:00:00   ITA               5                Veneto   
17996  2020-07-10T17:00:00   ITA               5                Veneto   
17997  2020-07-10T17:00:00   ITA               5                Veneto   
17998  2020-07-10T17:00:00   ITA               5                Veneto   
17999  2020-07-10T17:00:00   ITA               5                Veneto   

       codice_provincia               denominazione_provincia sigla_provincia  \
0                    66                              L'Aquila              AQ   
1                    67                                Teramo              TE   
2                    68                               Pescara              PE   
3                    69                                Chieti              CH   
4                   979  In fase di definizione/aggiornamento            None   
...                 ...                                   ...             ...   
17995                27                               Venezia              VE   
17996                28                                Padova              PD   
17997                29                                Rovigo              RO   
17998               899    Fuori Regione / Provincia Autonoma                   
17999               999                In fase di definizione                   

             lat       long  totale_casi  note  
0      42.351222  13.398438            0  None  
1      42.658918  13.704400            0  None  
2      42.464584  14.213648            0  None  
3      42.351032  14.167546            0  None  
4            NaN        NaN            0  None  
...          ...        ...          ...   ...  
17995  45.434905  12.338452         2696        
17996  45.406930  11.876087         3979        
17997  45.071073  11.790070          446        
17998        NaN        NaN          330        
17999        NaN        NaN           20        

[18000 rows x 11 columns]

data stato  codice_regione denominazione_regione  \
4      2020-02-24T18:00:00   ITA              13               Abruzzo   
7      2020-02-24T18:00:00   ITA              17            Basilicata   
13     2020-02-24T18:00:00   ITA              18              Calabria   
19     2020-02-24T18:00:00   ITA              15              Campania   
29     2020-02-24T18:00:00   ITA               8        Emilia-Romagna   
...                    ...   ...             ...                   ...   
15591  2020-06-24T17:00:00   ITA              19               Sicilia   
15602  2020-06-24T17:00:00   ITA               9               Toscana   
15605  2020-06-24T17:00:00   ITA              10                Umbria   
15607  2020-06-24T17:00:00   ITA               2         Valle d'Aosta   
15615  2020-06-24T17:00:00   ITA               5                Veneto   

       codice_provincia               denominazione_provincia sigla_provincia  \
4                   979  In fase di definizione/aggiornamento            None   
7                   980  In fase di definizione/aggiornamento            None   
13                  982  In fase di definizione/aggiornamento            None   
19                  983  In fase di definizione/aggiornamento            None   
29                  984  In fase di definizione/aggiornamento            None   
...                 ...                                   ...             ...   
15591               994  In fase di definizione/aggiornamento            None   
15602               995  In fase di definizione/aggiornamento            None   
15605               997  In fase di definizione/aggiornamento            None   
15607               998  In fase di definizione/aggiornamento            None   
15615               999  In fase di definizione/aggiornamento            None   

       lat  long  totale_casi  note  
4      NaN   NaN            0  None  
7      NaN   NaN            0  None  
13     NaN   NaN            0  None  
19     NaN   NaN            0  None  
29     NaN   NaN            0  None  
...    ...   ...          ...   ...  
15591  NaN   NaN           24  None  
15602  NaN   NaN          471  None  
15605  NaN   NaN           56  None  
15607  NaN   NaN            0  None  
15615  NaN   NaN          347  None  

[2562 rows x 11 columns]

Let's check all columns to clean them.  
Firts of all let's define a utility function

In [5]:
def show_distinct(p_dataframe, p_column_name):
    # In this way we can manage single columns (passed as string) 
    # and multiple columns passed as list
    if not isinstance(p_column_name, list):
        p_column_name = [p_column_name]
    display(p_dataframe[p_column_name].drop_duplicates().sort_values(by = p_column_name))

#### 2.2.1.1 Date
Here we don't have dirty data but date column can be converted to a Time Series to have correctly structured data

In [6]:
show_distinct(cities, 'data')

data
0      2020-02-24T18:00:00
128    2020-02-25T18:00:00
256    2020-02-26T18:00:00
384    2020-02-27T18:00:00
512    2020-02-28T18:00:00
...                    ...
17255  2020-07-06T17:00:00
17404  2020-07-07T17:00:00
17553  2020-07-08T17:00:00
17702  2020-07-09T17:00:00
17851  2020-07-10T17:00:00

[138 rows x 1 columns]

In [7]:
cities['data'] = pd.to_datetime(cities['data'], format = "%Y-%m-%dT%H:%M:%S")
display(cities[['data']].dtypes)

data
0     2020-02-24 18:00:00
128   2020-02-25 18:00:00
256   2020-02-26 18:00:00
384   2020-02-27 18:00:00
512   2020-02-28 18:00:00
...                   ...
17255 2020-07-06 17:00:00
17404 2020-07-07 17:00:00
17553 2020-07-08 17:00:00
17702 2020-07-09 17:00:00
17851 2020-07-10 17:00:00

[138 rows x 1 columns]

data    datetime64[ns]
dtype: object

#### 2.2.1.2 Country
Everything ok

In [8]:
show_distinct(cities, 'stato')

stato
0   ITA

#### 2.2.1.3 Region
There are some dirty data on Bolzano and Trento

In [30]:
show_distinct(cities, ['denominazione_regione', 'codice_regione'])

denominazione_regione  codice_regione
0                    Abruzzo              13
5                 Basilicata              17
8                   Calabria              18
14                  Campania              15
20            Emilia-Romagna               8
30     Friuli Venezia Giulia               6
35                     Lazio              12
41                   Liguria               7
46                 Lombardia               3
59                    Marche              11
65                    Molise              14
15695           P.A. Bolzano               4
68              P.A. Bolzano              21
15698            P.A. Trento               4
70               P.A. Trento              22
72                  Piemonte               1
81                    Puglia              16
88                  Sardegna              20
94                   Sicilia              19
104                  Toscana               9
115                   Umbria              10
118            Valle d'Aosta               2
120                   Veneto               5

It seems that there are 2 rows for the same day in Bolzano and Trento. The duplicated rows have dirty city name, null lat, null long, ... 
In following cells we will clean those data

In [ ]:
display(cities.loc[(cities.denominazione_regione == "P.A. Bolzano")\
                  ].sort_values(by = 'data').drop(columns = ['stato', \
                                                             'codice_regione', \
                                                             'denominazione_regione',\
                                                             'note']))
# Series with all dates of Bolzano "clean" data
bolzano_clean = cities.loc[(cities.denominazione_regione == "P.A. Bolzano") & \
                                   (cities.denominazione_provincia == "Bolzano"), 'data']

# Series with all dates of Bolzano "dirty" data
bolzano_dirty = cities.loc[(cities.denominazione_regione == "P.A. Bolzano") & \
                           (cities.denominazione_provincia != "Bolzano"), 'data']

Check that all dates of bolzano_dirty are included in bolzano_clean

In [ ]:
display(bolzano_dirty.subtract(bolzano_clean).drop_duplicates())

cities.drop(cities[(cities.denominazione_regione == "P.A. Bolzano") & \
                   (cities.denominazione_provincia != "Bolzano")].index, inplace = True)

Bolzano's data are splitted on 2 different region codes, likely an error, the right one is 4 (in common with Trento)

In [33]:
display(cities.groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])
cities.loc[cities.codice_regione == 21, 'codice_regione'] = 4
display(cities.loc[cities.denominazione_regione == "P.A. Bolzano"].groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])

data
denominazione_regione codice_regione      
Abruzzo               13               138
Basilicata            17               138
Calabria              18               138
Campania              15               138
Emilia-Romagna        8                138
Friuli Venezia Giulia 6                138
Lazio                 12               138
Liguria               7                138
Lombardia             3                138
Marche                11               138
Molise                14               138
P.A. Bolzano          4                 16
                      21               122
P.A. Trento           4                 16
                      22               122
Piemonte              1                138
Puglia                16               138
Sardegna              20               138
Sicilia               19               138
Toscana               9                138
Umbria                10               138
Valle d'Aosta         2                138
Veneto                5                138

data
denominazione_regione codice_regione      
P.A. Bolzano          4                138

Trento has the same duplicated data and wrong region code, we will do same operations as above (no comments)

In [38]:
display(cities.loc[(cities.denominazione_regione == "P.A. Trento")\
                  ].sort_values(by = 'data').drop(columns = ['stato', \
                                                             'codice_regione', \
                                                             'denominazione_regione',\
                                                             'note']))
trento_clean = cities.loc[(cities.denominazione_regione == "P.A. Trento") & \
                          (cities.denominazione_provincia == "Trento"), 'data']

trento_dirty = cities.loc[(cities.denominazione_regione == "P.A. Trento") & \
                          (cities.denominazione_provincia != "Trento"), 'data']

cities.drop(cities[(cities.denominazione_regione == "P.A. Trento") & \
                   (cities.denominazione_provincia != "Trento")].index, inplace = True)
cities.loc[cities.codice_regione == 22, 'codice_regione'] = 4

data  codice_provincia  \
70    2020-02-24 18:00:00                22   
71    2020-02-24 18:00:00               996   
198   2020-02-25 18:00:00                22   
199   2020-02-25 18:00:00               996   
326   2020-02-26 18:00:00                22   
...                   ...               ...   
17784 2020-07-09 17:00:00                22   
17786 2020-07-09 17:00:00               996   
17934 2020-07-10 17:00:00               896   
17933 2020-07-10 17:00:00                22   
17935 2020-07-10 17:00:00               996   

                    denominazione_provincia sigla_provincia        lat  \
70                                   Trento              TN  46.068935   
71     In fase di definizione/aggiornamento            None        NaN   
198                                  Trento              TN  46.068935   
199    In fase di definizione/aggiornamento            None        NaN   
326                                  Trento              TN  46.068935   
...                                     ...             ...        ...   
17784                                Trento              TN  46.068935   
17786                In fase di definizione                        NaN   
17934    Fuori Regione / Provincia Autonoma                        NaN   
17933                                Trento              TN  46.068935   
17935                In fase di definizione                        NaN   

            long  totale_casi  
70     11.121231            0  
71           NaN            0  
198    11.121231            0  
199          NaN            0  
326    11.121231            0  
...          ...          ...  
17784  11.121231         4876  
17786        NaN            0  
17934        NaN            0  
17933  11.121231         4879  
17935        NaN            0  

[292 rows x 7 columns]

Correct region name

In [40]:
cities.loc[cities.codice_regione == 4, 'denominazione_regione'] = "Trentino-Alto Adige"
display(cities.groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])

data
denominazione_regione codice_regione      
Abruzzo               13               138
Basilicata            17               138
Calabria              18               138
Campania              15               138
Emilia-Romagna        8                138
Friuli Venezia Giulia 6                138
Lazio                 12               138
Liguria               7                138
Lombardia             3                138
Marche                11               138
Molise                14               138
Piemonte              1                138
Puglia                16               138
Sardegna              20               138
Sicilia               19               138
Toscana               9                138
Trentino-Alto Adige   4                138
Umbria                10               138
Valle d'Aosta         2                138
Veneto                5                138

In [47]:
cities.loc[(cities.codice_regione == 13) & (cities.data == pd.to_datetime("2020-02-24 18:00:00"))]

data stato  codice_regione denominazione_regione  \
0 2020-02-24 18:00:00   ITA              13               Abruzzo   
1 2020-02-24 18:00:00   ITA              13               Abruzzo   
2 2020-02-24 18:00:00   ITA              13               Abruzzo   
3 2020-02-24 18:00:00   ITA              13               Abruzzo   
4 2020-02-24 18:00:00   ITA              13               Abruzzo   

   codice_provincia               denominazione_provincia sigla_provincia  \
0                66                              L'Aquila              AQ   
1                67                                Teramo              TE   
2                68                               Pescara              PE   
3                69                                Chieti              CH   
4               979  In fase di definizione/aggiornamento            None   

         lat       long  totale_casi  note  
0  42.351222  13.398438            0  None  
1  42.658918  13.704400            0  None  
2  42.464584  14.213648            0  None  
3  42.351032  14.167546            0  None  
4        NaN        NaN            0  None

#### 2.2.1.4 City
There are still some dirty rows, some regions seems to have too many cities, some "updating" rows.

In [63]:
display(cities.groupby(['denominazione_regione', 'codice_regione']).nunique()[['codice_provincia', 'denominazione_provincia']])
display(cities.loc[cities.denominazione_regione == 'Abruzzo'].groupby(['denominazione_regione', 'codice_regione', 'codice_provincia', 'denominazione_provincia']).nunique(['data']))

codice_provincia  \
denominazione_regione codice_regione                     
Abruzzo               13                             4   
Basilicata            17                             2   
Calabria              18                             5   
Campania              15                             5   
Emilia-Romagna        8                              9   
Friuli Venezia Giulia 6                              4   
Lazio                 12                             5   
Liguria               7                              4   
Lombardia             3                             12   
Marche                11                             5   
Molise                14                             2   
Piemonte              1                              8   
Puglia                16                             6   
Sardegna              20                             5   
Sicilia               19                             9   
Toscana               9                             10   
Trentino-Alto Adige   4                              2   
Umbria                10                             2   
Valle d'Aosta         2                              1   
Veneto                5                              7   

                                      denominazione_provincia  
denominazione_regione codice_regione                           
Abruzzo               13                                    4  
Basilicata            17                                    2  
Calabria              18                                    5  
Campania              15                                    5  
Emilia-Romagna        8                                     9  
Friuli Venezia Giulia 6                                     4  
Lazio                 12                                    5  
Liguria               7                                     4  
Lombardia             3                                    12  
Marche                11                                    5  
Molise                14                                    2  
Piemonte              1                                     8  
Puglia                16                                    6  
Sardegna              20                                    5  
Sicilia               19                                    9  
Toscana               9                                    10  
Trentino-Alto Adige   4                                     2  
Umbria                10                                    2  
Valle d'Aosta         2                                     1  
Veneto                5                                     7

data  \
denominazione_regione codice_regione codice_provincia denominazione_provincia         
Abruzzo               13             66               L'Aquila                  138   
                                     67               Teramo                    138   
                                     68               Pescara                   138   
                                     69               Chieti                    138   

                                                                               stato  \
denominazione_regione codice_regione codice_provincia denominazione_provincia          
Abruzzo               13             66               L'Aquila                     1   
                                     67               Teramo                       1   
                                     68               Pescara                      1   
                                     69               Chieti                       1   

                                                                               codice_regione  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                   
Abruzzo               13             66               L'Aquila                              1   
                                     67               Teramo                                1   
                                     68               Pescara                               1   
                                     69               Chieti                                1   

                                                                               denominazione_regione  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                          
Abruzzo               13             66               L'Aquila                                     1   
                                     67               Teramo                                       1   
                                     68               Pescara                                      1   
                                     69               Chieti                                       1   

                                                                               codice_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                     
Abruzzo               13             66               L'Aquila                                1   
                                     67               Teramo                                  1   
                                     68               Pescara                                 1   
                                     69               Chieti                                  1   

                                                                               denominazione_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                            
Abruzzo               13             66               L'Aquila                                       1   
                                     67               Teramo                                         1   
                                     68               Pescara                                        1   
                                     69               Chieti                                         1   

                                                                               sigla_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                    
Abruzzo               13             66               L'Aquila                               1   
                                     67               Teramo                                 1   
                                     68               Pescara                                1   
                                     69               Chieti                                 1   

The real cities of the region have the correct number of dates (rows), so we can drop those "dirty" rows

In [81]:
cities.drop(cities[cities['denominazione_provincia'].isin(('Fuori Regione / Provincia Autonoma',\
                                                           'In fase di definizione',\
                                                           'In fase di definizione/aggiornamento'))].index,\
            inplace = True)

display(cities.groupby(['denominazione_regione', 'codice_regione']).nunique()[['codice_provincia', 'denominazione_provincia']])

codice_provincia  \
denominazione_regione codice_regione                     
Abruzzo               13                             4   
Basilicata            17                             2   
Calabria              18                             5   
Campania              15                             5   
Emilia-Romagna        8                              9   
Friuli Venezia Giulia 6                              4   
Lazio                 12                             5   
Liguria               7                              4   
Lombardia             3                             12   
Marche                11                             5   
Molise                14                             2   
Piemonte              1                              8   
Puglia                16                             6   
Sardegna              20                             5   
Sicilia               19                             9   
Toscana               9                             10   
Trentino-Alto Adige   4                              2   
Umbria                10                             2   
Valle d'Aosta         2                              1   
Veneto                5                              7   

                                      denominazione_provincia  
denominazione_regione codice_regione                           
Abruzzo               13                                    4  
Basilicata            17                                    2  
Calabria              18                                    5  
Campania              15                                    5  
Emilia-Romagna        8                                     9  
Friuli Venezia Giulia 6                                     4  
Lazio                 12                                    5  
Liguria               7                                     4  
Lombardia             3                                    12  
Marche                11                                    5  
Molise                14                                    2  
Piemonte              1                                     8  
Puglia                16                                    6  
Sardegna              20                                    5  
Sicilia               19                                    9  
Toscana               9                                    10  
Trentino-Alto Adige   4                                     2  
Umbria                10                                    2  
Valle d'Aosta         2                                     1  
Veneto                5                                     7

There is also a null city short name, we can correct it

In [83]:
display(cities.loc[cities.sigla_provincia.astype(str).map(len) != 2, ['denominazione_provincia', 'sigla_provincia']])
cities.loc[(cities.sigla_provincia.astype(str).map(len) != 2) &\
           (cities.denominazione_provincia == 'Napoli'), 'sigla_provincia'] = "NA"
display(cities.loc[cities.sigla_provincia.astype(str).map(len) != 2, ['denominazione_provincia', 'sigla_provincia']])

Empty DataFrame
Columns: [denominazione_provincia, sigla_provincia]
Index: []

Empty DataFrame
Columns: [denominazione_provincia, sigla_provincia]
Index: []

#### 2.2.1.5 Position
Latitude and logintude with 0 value have already been cleaned in previous steps

In [85]:
display(cities[['lat']].drop_duplicates().sort_values(by = 'lat'))
display(cities[['long']].drop_duplicates().sort_values(by = 'long'))

lat
101  36.925092
102  37.059917
97   37.309711
98   37.492132
100  37.502878
..         ...
30   46.062555
70   46.068935
122  46.138375
48   46.170993
68   46.499335

[107 rows x 1 columns]

long
118   7.320149
75    7.551172
72    7.680687
41    8.027850
78    8.054082
..         ...
82   16.867367
11   17.125389
83   17.232372
84   17.946016
85   18.171897

[107 rows x 1 columns]

#### 2.2.1.6 Number of cases
Everything ok, no suspect values

In [91]:
display(cities[['totale_casi']].drop_duplicates().sort_values(by = 'totale_casi'))

totale_casi
0                0
156              1
307              2
151              3
149              4
...            ...
17312        24499
17461        24504
17610        24520
17759        24555
17908        24572

[3941 rows x 1 columns]

### 2.2.2 Region Dataframe

In [176]:
with open('Resources/dpc-covid19-ita-regioni.json') as f:
    d = json.load(f)
regions  = pd.DataFrame(d)    

After importing the data we can easily notice that there are some dirty data

In [177]:
regions

data stato  codice_regione denominazione_regione  \
0     2020-02-24T18:00:00   ITA              13               Abruzzo   
1     2020-02-24T18:00:00   ITA              17            Basilicata   
2     2020-02-24T18:00:00   ITA              18              Calabria   
3     2020-02-24T18:00:00   ITA              15              Campania   
4     2020-02-24T18:00:00   ITA               8        Emilia-Romagna   
...                   ...   ...             ...                   ...   
2893  2020-07-10T17:00:00   ITA              19               Sicilia   
2894  2020-07-10T17:00:00   ITA               9               Toscana   
2895  2020-07-10T17:00:00   ITA              10                Umbria   
2896  2020-07-10T17:00:00   ITA               2         Valle d'Aosta   
2897  2020-07-10T17:00:00   ITA               5                Veneto   

            lat       long  ricoverati_con_sintomi  terapia_intensiva  \
0     42.351222  13.398438                       0                  0   
1     40.639471  15.805148                       0                  0   
2     38.905976  16.594402                       0                  0   
3     40.839566  14.250850                       0                  0   
4     44.494367  11.341721                      10                  2   
...         ...        ...                     ...                ...   
2893  38.115697  13.362357                       6                  0   
2894  43.769231  11.255889                       9                  2   
2895  43.106758  12.388247                       4                  0   
2896  45.737503   7.320149                       4                  0   
2897  45.434905  12.338452                      14                  3   

      totale_ospedalizzati  isolamento_domiciliare  ...  \
0                        0                       0  ...   
1                        0                       0  ...   
2                        0                       0  ...   
3                        0                       0  ...   
4                       12                       6  ...   
...                    ...                     ...  ...   
2893                     6                     118  ...   
2894                    11                     319  ...   
2895                     4                       4  ...   
2896                     4                       1  ...   
2897                    17                     377  ...   

      variazione_totale_positivi  nuovi_positivi  dimessi_guariti  deceduti  \
0                              0               0                0         0   
1                              0               0                0         0   
2                              0               0                0         0   
3                              0               0                0         0   
4                              0              18                0         0   
...                          ...             ...              ...       ...   
2893                          -4               0             2691       283   
2894                          -7               2             8865      1121   
2895                           1               1             1360        80   
2896                           0               0             1045       146   
2897                           2              11            16936      2039   

      casi_da_sospetto_diagnostico  casi_da_screening  totale_casi  tamponi  \
0                              NaN                NaN            0        5   
1                              NaN                NaN            0        0   
2                              NaN                NaN            0        1   
3                              NaN                NaN            0       10   
4                              NaN                NaN           18      148   
...                            ...                ...          ...      ...   
2893                        2749.0              349.0   

#### 2.2.1.1 NaN
Some numeric fields have some NaN values

In [178]:
regions.iloc[:,6:19].fillna(value = 0, inplace = True)

#### 2.2.1.2 Date
Here we don't have dirty data but date column can be converted to a Time Series to have correctly structured data

In [179]:
show_distinct(regions, 'data')

data
0     2020-02-24T18:00:00
21    2020-02-25T18:00:00
42    2020-02-26T18:00:00
63    2020-02-27T18:00:00
84    2020-02-28T18:00:00
...                   ...
2793  2020-07-06T17:00:00
2814  2020-07-07T17:00:00
2835  2020-07-08T17:00:00
2856  2020-07-09T17:00:00
2877  2020-07-10T17:00:00

[138 rows x 1 columns]

In [180]:
regions['data'] = pd.to_datetime(regions['data'], format = "%Y-%m-%dT%H:%M:%S")
display(regions[['data']].dtypes)

data    datetime64[ns]
dtype: object

#### 2.2.1.3 Country
Everything ok

In [181]:
show_distinct(regions, 'stato')

stato
0   ITA

#### 2.2.1.4 Region
Same problem of City dataframe with "dirty" region name and code on Trento and Bolzano. In this case we have to:
  - Update Bolzano and Trento region name and region code.
  - Overwrite lat and long of Bolzano rows with info from Trento. 
    In all other rows lat and long point to capital of the region
  - Aggregate Bolzano and Trento data in a single row for each day

In [182]:
type(regions.loc[regions.codice_regione == 22, ['lat']].iloc[0, 0])

numpy.float64

In [183]:
before_styler = regions[['denominazione_regione', 'codice_regione']].drop_duplicates().sort_values(by = ['denominazione_regione', 'codice_regione']).style.set_table_attributes("style='display:inline'").set_caption("Before")
regions.loc[regions.codice_regione == 21, 'lat'] = regions.loc[regions.codice_regione == 22, ['lat']].iloc[0, 0]
regions.loc[regions.codice_regione == 21, 'long'] = regions.loc[regions.codice_regione == 22, ['long']].iloc[0, 0]
regions.loc[regions.codice_regione == 21, 'codice_regione'] = 4
regions.loc[regions.codice_regione == 22, 'codice_regione'] = 4
regions.loc[regions.codice_regione == 4, 'denominazione_regione'] = "Trentino-Alto Adige"
after_styler = regions[['denominazione_regione', 'codice_regione']].drop_duplicates().sort_values(by = ['denominazione_regione', 'codice_regione']).style.set_table_attributes("style='display:inline'").set_caption("After")
display_html(before_styler._repr_html_() + " " + after_styler._repr_html_(), raw=True)

Before denominazione_regione codice_regione 
 
 0 
 Abruzzo 
 13 
 
 
 1 
 Basilicata 
 17 
 
 
 2 
 Calabria 
 18 
 
 
 3 
 Campania 
 15 
 
 
 4 
 Emilia-Romagna 
 8 
 
 
 5 
 Friuli Venezia Giulia 
 6 
 
 
 6 
 Lazio 
 12 
 
 
 7 
 Liguria 
 7 
 
 
 8 
 Lombardia 
 3 
 
 
 9 
 Marche 
 11 
 
 
 10 
 Molise 
 14 
 
 
 11 
 P.A. Bolzano 
 21 
 
 
 12 
 P.A. Trento 
 22 
 
 
 13 
 Piemonte 
 1 
 
 
 14 
 Puglia 
 16 
 
 
 15 
 Sardegna 
 20 
 
 
 16 
 Sicilia 
 19 
 
 
 17 
 Toscana 
 9 
 
 
 18 
 Umbria 
 10 
 
 
 19 
 Valle d'Aosta 
 2 
 
 
 20 
 Veneto 
 5 
 
 After denominazione_regione codice_regione 
 
 0 
 Abruzzo 
 13 
 
 
 1 
 Basilicata 
 17 
 
 
 2 
 Calabria 
 18 
 
 
 3 
 Campania 
 15 
 
 
 4 
 Emilia-Romagna 
 8 
 
 
 5 
 Friuli Venezia Giulia 
 6 
 
 
 6 
 Lazio 
 12 
 
 
 7 
 Liguria 
 7 
 
 
 8 
 Lombardia 
 3 
 
 
 9 
 Marche 
 11 
 
 
 10 
 Molise 
 14 
 
 
 13 
 Piemonte 
 1 
 
 
 14 
 Puglia 
 16 
 
 
 15 
 Sardegna 
 20 
 
 
 16 
 Sicilia 
 19 
 
 
 17 
 Toscana 
 9 
 
 
 11 
 Trentino-Alto Adige 
 4 
 
 
 18 
 Umbria 
 10 
 
 
 19 
 Valle d'Aosta 
 2 
 
 
 20 
 Veneto 
 5

After merging Bolzano and Trento in the same region we have to aggregate their data grouping by date

In [191]:
regions.loc[regions.codice_regione == 4].groupby(['data', 'stato', 'codice_regione', 'denominazione_regione', 'lat', 'long']).sum()

ricoverati_con_sintomi  \
data                stato codice_regione denominazione_regione lat       long                                
2020-02-24 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       0   
2020-02-25 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       1   
2020-02-26 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       1   
2020-02-27 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       1   
2020-02-28 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       1   
...                                                                                                    ...   
2020-07-06 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       6   
2020-07-07 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       6   
2020-07-08 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       9   
2020-07-09 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       7   
2020-07-10 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                       7   

                                                                                    terapia_intensiva  \
data                stato codice_regione denominazione_regione lat       long                           
2020-02-24 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-02-25 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-02-26 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-02-27 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-02-28 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
...                                                                                               ...   
2020-07-06 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-07-07 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-07-08 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-07-09 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   
2020-07-10 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                  0   

                                                                                    totale_ospedalizzati  \
data                stato codice_regione denominazione_regione lat       long                              
2020-02-24 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     0   
2020-02-25 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     1   
2020-02-26 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     1   
2020-02-27 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     1   
2020-02-28 18:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     1   
...                                                                                                  ...   
2020-07-06 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     6   
2020-07-07 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     6   
2020-07-08 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                     9   
2020-07-09 17:00:00 ITA   4              Trentino-Alto Adige   46.068935 11.121231                 

Bolzano's data are splitted on 2 different region codes, likely an error, the right one is 4 (in common with Trento)

In [33]:
display(regions.groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])

display(regions.loc[regions.denominazione_regione == "P.A. Bolzano"].groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])

data
denominazione_regione codice_regione      
Abruzzo               13               138
Basilicata            17               138
Calabria              18               138
Campania              15               138
Emilia-Romagna        8                138
Friuli Venezia Giulia 6                138
Lazio                 12               138
Liguria               7                138
Lombardia             3                138
Marche                11               138
Molise                14               138
P.A. Bolzano          4                 16
                      21               122
P.A. Trento           4                 16
                      22               122
Piemonte              1                138
Puglia                16               138
Sardegna              20               138
Sicilia               19               138
Toscana               9                138
Umbria                10               138
Valle d'Aosta         2                138
Veneto                5                138

data
denominazione_regione codice_regione      
P.A. Bolzano          4                138

Trento has the same duplicated data and wrong region code, we will do same operations as above (no comments)

In [38]:
display(regions.loc[(regions.denominazione_regione == "P.A. Trento")\
                  ].sort_values(by = 'data').drop(columns = ['stato', \
                                                             'codice_regione', \
                                                             'denominazione_regione',\
                                                             'note']))
trento_clean = regions.loc[(regions.denominazione_regione == "P.A. Trento") & \
                          (regions.denominazione_provincia == "Trento"), 'data']

trento_dirty = regions.loc[(regions.denominazione_regione == "P.A. Trento") & \
                          (regions.denominazione_provincia != "Trento"), 'data']

regions.drop(regions[(regions.denominazione_regione == "P.A. Trento") & \
                   (regions.denominazione_provincia != "Trento")].index, inplace = True)
regions.loc[regions.codice_regione == 22, 'codice_regione'] = 4

data  codice_provincia  \
70    2020-02-24 18:00:00                22   
71    2020-02-24 18:00:00               996   
198   2020-02-25 18:00:00                22   
199   2020-02-25 18:00:00               996   
326   2020-02-26 18:00:00                22   
...                   ...               ...   
17784 2020-07-09 17:00:00                22   
17786 2020-07-09 17:00:00               996   
17934 2020-07-10 17:00:00               896   
17933 2020-07-10 17:00:00                22   
17935 2020-07-10 17:00:00               996   

                    denominazione_provincia sigla_provincia        lat  \
70                                   Trento              TN  46.068935   
71     In fase di definizione/aggiornamento            None        NaN   
198                                  Trento              TN  46.068935   
199    In fase di definizione/aggiornamento            None        NaN   
326                                  Trento              TN  46.068935   
...                                     ...             ...        ...   
17784                                Trento              TN  46.068935   
17786                In fase di definizione                        NaN   
17934    Fuori Regione / Provincia Autonoma                        NaN   
17933                                Trento              TN  46.068935   
17935                In fase di definizione                        NaN   

            long  totale_casi  
70     11.121231            0  
71           NaN            0  
198    11.121231            0  
199          NaN            0  
326    11.121231            0  
...          ...          ...  
17784  11.121231         4876  
17786        NaN            0  
17934        NaN            0  
17933  11.121231         4879  
17935        NaN            0  

[292 rows x 7 columns]

Correct region name

In [40]:

display(regions.groupby(['denominazione_regione', 'codice_regione']).nunique()[['data']])

data
denominazione_regione codice_regione      
Abruzzo               13               138
Basilicata            17               138
Calabria              18               138
Campania              15               138
Emilia-Romagna        8                138
Friuli Venezia Giulia 6                138
Lazio                 12               138
Liguria               7                138
Lombardia             3                138
Marche                11               138
Molise                14               138
Piemonte              1                138
Puglia                16               138
Sardegna              20               138
Sicilia               19               138
Toscana               9                138
Trentino-Alto Adige   4                138
Umbria                10               138
Valle d'Aosta         2                138
Veneto                5                138

In [47]:
regions.loc[(regions.codice_regione == 13) & (regions.data == pd.to_datetime("2020-02-24 18:00:00"))]

data stato  codice_regione denominazione_regione  \
0 2020-02-24 18:00:00   ITA              13               Abruzzo   
1 2020-02-24 18:00:00   ITA              13               Abruzzo   
2 2020-02-24 18:00:00   ITA              13               Abruzzo   
3 2020-02-24 18:00:00   ITA              13               Abruzzo   
4 2020-02-24 18:00:00   ITA              13               Abruzzo   

   codice_provincia               denominazione_provincia sigla_provincia  \
0                66                              L'Aquila              AQ   
1                67                                Teramo              TE   
2                68                               Pescara              PE   
3                69                                Chieti              CH   
4               979  In fase di definizione/aggiornamento            None   

         lat       long  totale_casi  note  
0  42.351222  13.398438            0  None  
1  42.658918  13.704400            0  None  
2  42.464584  14.213648            0  None  
3  42.351032  14.167546            0  None  
4        NaN        NaN            0  None

#### 2.2.1.4 City
There are still some dirty rows, some regions seems to have too many regions, some "updating" rows.

In [63]:
display(regions.groupby(['denominazione_regione', 'codice_regione']).nunique()[['codice_provincia', 'denominazione_provincia']])
display(regions.loc[regions.denominazione_regione == 'Abruzzo'].groupby(['denominazione_regione', 'codice_regione', 'codice_provincia', 'denominazione_provincia']).nunique(['data']))

codice_provincia  \
denominazione_regione codice_regione                     
Abruzzo               13                             4   
Basilicata            17                             2   
Calabria              18                             5   
Campania              15                             5   
Emilia-Romagna        8                              9   
Friuli Venezia Giulia 6                              4   
Lazio                 12                             5   
Liguria               7                              4   
Lombardia             3                             12   
Marche                11                             5   
Molise                14                             2   
Piemonte              1                              8   
Puglia                16                             6   
Sardegna              20                             5   
Sicilia               19                             9   
Toscana               9                             10   
Trentino-Alto Adige   4                              2   
Umbria                10                             2   
Valle d'Aosta         2                              1   
Veneto                5                              7   

                                      denominazione_provincia  
denominazione_regione codice_regione                           
Abruzzo               13                                    4  
Basilicata            17                                    2  
Calabria              18                                    5  
Campania              15                                    5  
Emilia-Romagna        8                                     9  
Friuli Venezia Giulia 6                                     4  
Lazio                 12                                    5  
Liguria               7                                     4  
Lombardia             3                                    12  
Marche                11                                    5  
Molise                14                                    2  
Piemonte              1                                     8  
Puglia                16                                    6  
Sardegna              20                                    5  
Sicilia               19                                    9  
Toscana               9                                    10  
Trentino-Alto Adige   4                                     2  
Umbria                10                                    2  
Valle d'Aosta         2                                     1  
Veneto                5                                     7

data  \
denominazione_regione codice_regione codice_provincia denominazione_provincia         
Abruzzo               13             66               L'Aquila                  138   
                                     67               Teramo                    138   
                                     68               Pescara                   138   
                                     69               Chieti                    138   

                                                                               stato  \
denominazione_regione codice_regione codice_provincia denominazione_provincia          
Abruzzo               13             66               L'Aquila                     1   
                                     67               Teramo                       1   
                                     68               Pescara                      1   
                                     69               Chieti                       1   

                                                                               codice_regione  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                   
Abruzzo               13             66               L'Aquila                              1   
                                     67               Teramo                                1   
                                     68               Pescara                               1   
                                     69               Chieti                                1   

                                                                               denominazione_regione  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                          
Abruzzo               13             66               L'Aquila                                     1   
                                     67               Teramo                                       1   
                                     68               Pescara                                      1   
                                     69               Chieti                                       1   

                                                                               codice_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                     
Abruzzo               13             66               L'Aquila                                1   
                                     67               Teramo                                  1   
                                     68               Pescara                                 1   
                                     69               Chieti                                  1   

                                                                               denominazione_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                            
Abruzzo               13             66               L'Aquila                                       1   
                                     67               Teramo                                         1   
                                     68               Pescara                                        1   
                                     69               Chieti                                         1   

                                                                               sigla_provincia  \
denominazione_regione codice_regione codice_provincia denominazione_provincia                    
Abruzzo               13             66               L'Aquila                               1   
                                     67               Teramo                                 1   
                                     68               Pescara                                1   
                                     69               Chieti                                 1   

The real regions of the region have the correct number of dates (rows), so we can drop those "dirty" rows

In [81]:
regions.drop(regions[regions['denominazione_provincia'].isin(('Fuori Regione / Provincia Autonoma',\
                                                           'In fase di definizione',\
                                                           'In fase di definizione/aggiornamento'))].index,\
            inplace = True)

display(regions.groupby(['denominazione_regione', 'codice_regione']).nunique()[['codice_provincia', 'denominazione_provincia']])

codice_provincia  \
denominazione_regione codice_regione                     
Abruzzo               13                             4   
Basilicata            17                             2   
Calabria              18                             5   
Campania              15                             5   
Emilia-Romagna        8                              9   
Friuli Venezia Giulia 6                              4   
Lazio                 12                             5   
Liguria               7                              4   
Lombardia             3                             12   
Marche                11                             5   
Molise                14                             2   
Piemonte              1                              8   
Puglia                16                             6   
Sardegna              20                             5   
Sicilia               19                             9   
Toscana               9                             10   
Trentino-Alto Adige   4                              2   
Umbria                10                             2   
Valle d'Aosta         2                              1   
Veneto                5                              7   

                                      denominazione_provincia  
denominazione_regione codice_regione                           
Abruzzo               13                                    4  
Basilicata            17                                    2  
Calabria              18                                    5  
Campania              15                                    5  
Emilia-Romagna        8                                     9  
Friuli Venezia Giulia 6                                     4  
Lazio                 12                                    5  
Liguria               7                                     4  
Lombardia             3                                    12  
Marche                11                                    5  
Molise                14                                    2  
Piemonte              1                                     8  
Puglia                16                                    6  
Sardegna              20                                    5  
Sicilia               19                                    9  
Toscana               9                                    10  
Trentino-Alto Adige   4                                     2  
Umbria                10                                    2  
Valle d'Aosta         2                                     1  
Veneto                5                                     7

There is also a null city short name, we can correct it

In [83]:
display(regions.loc[regions.sigla_provincia.astype(str).map(len) != 2, ['denominazione_provincia', 'sigla_provincia']])
regions.loc[(regions.sigla_provincia.astype(str).map(len) != 2) &\
           (regions.denominazione_provincia == 'Napoli'), 'sigla_provincia'] = "NA"
display(regions.loc[regions.sigla_provincia.astype(str).map(len) != 2, ['denominazione_provincia', 'sigla_provincia']])

Empty DataFrame
Columns: [denominazione_provincia, sigla_provincia]
Index: []

Empty DataFrame
Columns: [denominazione_provincia, sigla_provincia]
Index: []

#### 2.2.1.5 Position
Latitude and logintude with 0 value have already been cleaned in previous steps

In [85]:
display(regions[['lat']].drop_duplicates().sort_values(by = 'lat'))
display(regions[['long']].drop_duplicates().sort_values(by = 'long'))

lat
101  36.925092
102  37.059917
97   37.309711
98   37.492132
100  37.502878
..         ...
30   46.062555
70   46.068935
122  46.138375
48   46.170993
68   46.499335

[107 rows x 1 columns]

long
118   7.320149
75    7.551172
72    7.680687
41    8.027850
78    8.054082
..         ...
82   16.867367
11   17.125389
83   17.232372
84   17.946016
85   18.171897

[107 rows x 1 columns]

regions[['denominazione_provincia', 'codice_provincia']].drop_duplicates().groupby(['denominazione_provincia']).filter(lambda x: len(x) > 1)

In [31]:
show_distinct(regions, ['denominazione_regione', 'codice_regione'])

#bolzano_dataframe.loc[bolzano_dataframe.denominazione_provincia != "Bolzano"][['data']].substract(bolzano_dataframe.loc[bolzano_dataframe.denominazione_provincia == "Bolzano"][['data']])
#regions.loc[regions.codice_provincia == 21][['data']].subtract(regions.loc[regions.codice_provincia == ][['data']])
#regions.loc[regions.denominazione_regione == "P.A. Bolzano"].sort_values(by = 'data')
#(regions.loc[regions.denominazione_regione == "P.A. Bolzano"].loc[regions.denominazione_provincia != "Bolzano"])#[['codice_provincia']].drop_duplicates()

data  codice_provincia denominazione_provincia  \
68    2020-02-24 18:00:00                21                 Bolzano   
196   2020-02-25 18:00:00                21                 Bolzano   
324   2020-02-26 18:00:00                21                 Bolzano   
452   2020-02-27 18:00:00                21                 Bolzano   
580   2020-02-28 18:00:00                21                 Bolzano   
...                   ...               ...                     ...   
17334 2020-07-06 17:00:00                21                 Bolzano   
17483 2020-07-07 17:00:00                21                 Bolzano   
17632 2020-07-08 17:00:00                21                 Bolzano   
17781 2020-07-09 17:00:00                21                 Bolzano   
17930 2020-07-10 17:00:00                21                 Bolzano   

      sigla_provincia        lat       long  totale_casi  
68                 BZ  46.499335  11.356624            0  
196                BZ  46.499335  11.356624            1  
324                BZ  46.499335  11.356624            1  
452                BZ  46.499335  11.356624            1  
580                BZ  46.499335  11.356624            1  
...               ...        ...        ...          ...  
17334              BZ  46.499335  11.356624         2647  
17483              BZ  46.499335  11.356624         2648  
17632              BZ  46.499335  11.356624         2650  
17781              BZ  46.499335  11.356624         2653  
17930              BZ  46.499335  11.356624         2658  

[138 rows x 7 columns]

68   NaT
Name: data, dtype: timedelta64[ns]

denominazione_regione  codice_regione
0                    Abruzzo              13
5                 Basilicata              17
8                   Calabria              18
14                  Campania              15
20            Emilia-Romagna               8
30     Friuli Venezia Giulia               6
35                     Lazio              12
41                   Liguria               7
46                 Lombardia               3
59                    Marche              11
65                    Molise              14
15695           P.A. Bolzano               4
68              P.A. Bolzano              21
15698            P.A. Trento               4
70               P.A. Trento              22
72                  Piemonte               1
81                    Puglia              16
88                  Sardegna              20
94                   Sicilia              19
104                  Toscana               9
115                   Umbria              10
118            Valle d'Aosta               2
120                   Veneto               5

In [13]:
regions.loc[(regions.denominazione_regione == "P.A. Bolzano") & \
           (regions.codice_regione == 21)]

data stato  codice_regione denominazione_regione  \
68    2020-02-24 18:00:00   ITA              21          P.A. Bolzano   
196   2020-02-25 18:00:00   ITA              21          P.A. Bolzano   
324   2020-02-26 18:00:00   ITA              21          P.A. Bolzano   
452   2020-02-27 18:00:00   ITA              21          P.A. Bolzano   
580   2020-02-28 18:00:00   ITA              21          P.A. Bolzano   
...                   ...   ...             ...                   ...   
15044 2020-06-20 17:00:00   ITA              21          P.A. Bolzano   
15172 2020-06-21 17:00:00   ITA              21          P.A. Bolzano   
15300 2020-06-22 17:00:00   ITA              21          P.A. Bolzano   
15428 2020-06-23 17:00:00   ITA              21          P.A. Bolzano   
15556 2020-06-24 17:00:00   ITA              21          P.A. Bolzano   

       codice_provincia denominazione_provincia sigla_provincia        lat  \
68                   21                 Bolzano              BZ  46.499335   
196                  21                 Bolzano              BZ  46.499335   
324                  21                 Bolzano              BZ  46.499335   
452                  21                 Bolzano              BZ  46.499335   
580                  21                 Bolzano              BZ  46.499335   
...                 ...                     ...             ...        ...   
15044                21                 Bolzano              BZ  46.499335   
15172                21                 Bolzano              BZ  46.499335   
15300                21                 Bolzano              BZ  46.499335   
15428                21                 Bolzano              BZ  46.499335   
15556                21                 Bolzano              BZ  46.499335   

            long  totale_casi  note  
68     11.356624            0  None  
196    11.356624            1  None  
324    11.356624            1  None  
452    11.356624            1  None  
580    11.356624            1  None  
...          ...          ...   ...  
15044  11.356624         2618  None  
15172  11.356624         2622  None  
15300  11.356624         2633  None  
15428  11.356624         2633  None  
15556  11.356624         2634  None  

[122 rows x 11 columns]

merged = pd.merge(dirty_rows, cities, on = 'codice_provincia')
display(merged.loc[merged.denominazione_provincia_y != "In fase di definizione/aggiornamento"])
#cities.drop(cities[(cities['lat'] == 0) | \
#                                   (cities['sigla_provincia'] == 0) | \
#                                   (cities['denominazione_provincia'] == 'In fase di definizione/aggiornamento') \
#                                  ].index, inplace = True)
#cities = (cities.loc[cities.codice_provincia == 979])[['denominazione_provincia', 'sigla_provincia']].drop_duplicates()
#display(cities)

---
# 3. Describe data 
Describe data using Pandas features and provide some plots according to your intuition (start small following the examples that we had done together); also, feel free to explore some connections you find interesting.
  - Provide visualizations such as plots, box-plots and scatter matrices (have a look at https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html  for inspiration).
  - The more combination you explore the better, in the sense to combine in the same figure multiple subplots or data series.

from ing_theme_matplotlib import mpl_style
mpl_style(dark = True)

---
# 4. Hierarchical indexes
The two files are correlated in the sense that the second adds details to the first. Figure out the connections and build new Pandas dataframes that use hierarchy indexes. Also perform some interesting marginalizations you like to report.
  - For instance, we expect to see a dataframe that is indexed by an outermost level of regions and an inner level of provinces; accordingly, the rows should report data with respect to this index.

The more graphics, descriptions and Pandas manipulation you experiment the better.